app/models/work_package/pdf_export/overview_table.rb

```diff
@@ -0,0 +1,178 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module WorkPackage::PDFExport::OverviewTable
+  def write_work_packages_overview!(work_packages, id_wp_meta_map)
+    if query.grouped?
+      write_grouped!(work_packages, id_wp_meta_map)
+    else
+      with_margin(overview_table_margins_style) do
+        write_table!(work_packages, id_wp_meta_map, get_total_sums)
+      end
+    end
+  end
+
+  private
+
+  def write_grouped!(work_packages, id_wp_meta_map)
+    groups = {}
+    work_packages.each do |work_package|
+      group = query.group_by_column.value(work_package)
+      groups[group] = groups[group] || []
+      groups[group].push work_package
+    end
```

Comment:
:recycle: Ruby has [group_by](https://ruby-doc.org/3.2.2/Enumerable.html#method-i-group_by) on enumerables:

```suggestion
    work_packages.group_by do |work_package|
      query.group_by_column.value(work_package)
    end
```

Reply:
:memo: Also, when initializing a Hash, you can define a default value for each unknown/new index:

```ruby
groups = Hash.new { |hash, key| hash[key] = Array.new }
```

With this you can just use `groups[group] << work_package` and don't need to initialize the array every time

Reply:
Awesome, thanks!